### PR TITLE
LPS-116829 Support keywords from multiple countries in traffic sources

### PR DESCRIPTION
**Pull request by darquesdev**
**Motivation**

Traffic sources must support that the keyword list is breaked downs by countries. In addition, for each country, the country code and the country name must be provided in the user's language.

**Provided solution**

To carry out this story the DXP code has been adapted to accept the AC endpoint that is already serving this information.

**How to test it**

A connection with AC is required, the new code is already in the testing environment:

See endpoint [http://192.168.40.21:8081/api/seo/1.0/traffic-sources?url=http://myliferay1:8080/-/the-pragmatic-programmer](http://192.168.40.21:8081/api/seo/1.0/traffic-sources?url=http://myliferay1:8080/-/the-pragmatic-programmer)

It's been impossible to validate that a list of keywords for a given country and URL, since this URL has to be previously indexed in Google.

Unit tests have been added where you can see the appearance of the new API provided by the front.

New JSON:
```javascript
[
    {
        "countryKeywords": [
            {
                "countryCode": "us",
                "countryName": "United States",
                "keywords":[
                    {
                        "keyword": "amazon",
                        "position": 1,
                        "searchVolume": 101000000,
                        "traffic": 80800000
                    },
                    {
                        "keyword": "amazon com",
                        "position": 1,
                        "searchVolume": 4090000,
                        "traffic": 3272000
                    }
                ]
            },
            {
                "countryCode": "es",
                "countryName": "Spain",
                "keywords":[
                    {
                        "keyword": "amazon",
                        "position": 1,
                        "searchVolume": 858440,
                        "traffic": 50000
                    },
                    {
                        "keyword": "amazon america",
                        "position": 1,
                        "searchVolume": 5400,
                        "traffic": 45000
                    }
                ]
            }
        ],
        "name": "paid",
        "share": 50,
        "title": "Paid",
        "value": 99999999,
        "helpMessage": "This number refers to the volume of people that find your page through paid keywords."
    },
    {
        "countryKeywords": [
            {
                "countryCode": "us",
                "countryName": "United States",
                "keywords":[
                    {
                        "keyword": "amazon",
                        "position": 1,
                        "searchVolume": 101000000,
                        "traffic": 80800000
                    },
                    {
                         "keyword": "amazon com",
                        "position": 1,
                        "searchVolume": 4090000,
                        "traffic": 3272000
                    }
                ]
            },
            {
                "countryCode": "es",
                "countryName": "Spain",
                "keywords":[
                    {
                        "keyword": "amazon",
                        "position": 1,
                        "searchVolume": 858440,
                        "traffic": 50000
                    },
                    {
                        "keyword": "amazon america",
                        "position": 1,
                        "searchVolume": 5400,
                        "traffic": 45000
                    }
                ]
            }
        ],        "name": "organic",
        "share": 50,
        "title": "Organic",
        "value": 99999999,
        "helpMessage": "This number refers to the volume of people that find your page through a search engine."
    }
]```


Old json:

```json[
    {
        "keywords": [
            {
                "keyword": "amazon",
                "position": 1,
                "searchVolume": 101000000,
                "traffic": 80800000
            },
            {
                "keyword": "amazon com",
                "position": 1,
                "searchVolume": 4090000,
                "traffic": 3272000
            }
        ],
        "name": "paid",
        "share": 50,
        "title": "Paid",
        "value": 99999999,
        "helpMessage": "This number refers to the volume of people that find your page through paid keywords."
    },
    {
        "keywords": [
            {
                "keyword": "amazon",
                "position": 1,
                "searchVolume": 101000000,
                "traffic": 80800000
            },
            {
                "keyword": "amazon com",
                "position": 1,
                "searchVolume": 4090000,
                "traffic": 3272000
            }
        ],
        "name": "organic",
        "share": 50,
        "title": "Organic",
        "value": 99999999,
        "helpMessage": "This number refers to the volume of people that find your page through a search engine."
    }
]```

### CODE REVIEW

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/display/context/AnalyticsReportsDisplayContext.java [hunk 1]
```diff
@@ -296,7 +296,8 @@ private JSONArray _getTrafficSourcesJSONArray() {
 					).findFirst(
 					).map(
 						trafficSource -> trafficSource.toJSONObject(
-							helpMessageMap.get(name), title)
+							helpMessageMap.get(name), _themeDisplay.getLocale(),
```
Comment (cgoncas) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/display/context/AnalyticsReportsDisplayContext.java:299]: Are we sure we want the user language? Or do we prefer the site default language or something like that?

Comment (darquesdev) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/display/context/AnalyticsReportsDisplayContext.java:299]: we need the current user locale, in order to display the country name using this language

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/model/CountrySearchKeywords.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+			return JSONFactoryUtil.createJSONArray();
+		}
+
+		Stream<SearchKeyword> stream = _searchKeywords.stream();
+
+		return JSONUtil.putAll(
+			stream.map(
+				SearchKeyword::toJSONObject
+			).limit(
+				5
```
Comment (cgoncas) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/model/CountrySearchKeywords.java:116]: Should this limit be configurable?

Comment (darquesdev) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/model/CountrySearchKeywords.java:116]: Good question, It's not a requirement in the story but I'm going to make it sure

**Comment by cgoncas** [comment]
ci:test:relevant

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/display/context/AnalyticsReportsDisplayContext.java (hunk 1 above)
Comment (cgoncas) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/display/context/AnalyticsReportsDisplayContext.java:299]: But are we retrieving the keywords in the user language?

**Comment by cgoncas** [comment]
ci:forward

**Pull request closed without merging**